File: appengine/monorail/features/federated.py
# Copyright 2019 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file or at
# https://developers.google.com/open-source/licenses/bsd

"""Logic for storing and representing issues from external trackers."""

from __future__ import print_function
from __future__ import division
from __future__ import absolute_import

import re

from framework.exceptions import InvalidExternalIssueReference


class FederatedIssue(object):
  """Abstract base class for holding one federated issue.

  Each distinct external tracker should subclass this.
  """
  shortlink_re = None

  def __init__(self, shortlink):
    if not self.IsShortlinkValid(shortlink):
      raise InvalidExternalIssueReference(
          'Shortlink does not match any valid tracker: %s' % shortlink)

    self.shortlink = shortlink

  @classmethod
  def IsShortlinkValid(cls, shortlink):
    """Returns whether given shortlink is correctly formatted."""
    if not cls.shortlink_re:
      raise NotImplementedError()
    return re.match(cls.shortlink_re, shortlink)


class GoogleIssueTrackerIssue(FederatedIssue):
  """Holds one Google Issue Tracker issue.

  URL: https://issuetracker.google.com/
  """
  shortlink_re = r'^b\/\d+$'
  url_format = 'https://issuetracker.google.com/issues/{issue_id}'

  def __init__(self, shortlink):
    super(GoogleIssueTrackerIssue, self).__init__(shortlink)
    self.issue_id = int(self.shortlink[2:])

  def ToURL(self):
    return self.url_format.format(issue_id=self.issue_id)

  def Summary(self):
    """Returns a short string description for UI."""
    return 'Google Issue Tracker issue %s.' % self.issue_id


# All supported tracker classes.
_federated_issue_classes = [GoogleIssueTrackerIssue]


def IsShortlinkValid(shortlink):
  """Returns whether the given string is valid for any issue tracker."""
  return any(tracker_class.IsShortlinkValid(shortlink)
      for tracker_class in _federated_issue_classes)


def FromShortlink(shortlink):
  """Returns a FederatedIssue for the first matching tracker.

  If no matching tracker is found, returns None.
  """
  for tracker_class in _federated_issue_classes:
    if tracker_class.IsShortlinkValid(shortlink):
      return tracker_class(shortlink)

  return None

File: appengine/monorail/static_src/shared/federated.js
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/**
 * Logic for dealing with federated issue references.
 */

import loadGapi from './gapi-loader';

const GOOGLE_ISSUE_TRACKER_REGEX = /^b\/\d+$/;

const GOOGLE_ISSUE_TRACKER_API_ROOT = 'https://issuetracker.corp.googleapis.com';
const GOOGLE_ISSUE_TRACKER_DISCOVERY_PATH = '/$discovery/rest';
const GOOGLE_ISSUE_TRACKER_API_VERSION = 'v1';

// Returns if shortlink is valid for any federated tracker.
export function isShortlinkValid(shortlink) {
  return FEDERATED_TRACKERS.some((TrackerClass) => {
    try {
      return new TrackerClass(shortlink);
    } catch (e) {
      if (e instanceof FederatedIssueError) {
        return false;
      } else {
        throw e;
      }
    }
  });
}

// Returns a issue instance for the first matching tracker.
export function fromShortlink(shortlink) {
  for (const key in FEDERATED_TRACKERS) {
    if (FEDERATED_TRACKERS.hasOwnProperty(key)) {
      const TrackerClass = FEDERATED_TRACKERS[key];
      try {
        return new TrackerClass(shortlink);
      } catch (e) {
        if (e instanceof FederatedIssueError) {
          continue;
        } else {
          throw e;
        }
      }
    }
  }
  return null;
}

// FederatedIssue is an abstract class for representing one federated issue.
// Each supported tracker should subclass this class.
class FederatedIssue {
  constructor(shortlink) {
    if (!this.isShortlinkValid(shortlink)) {
      throw new FederatedIssueError(`Invalid tracker shortlink: ${shortlink}`);
    }
    this.shortlink = shortlink;
  }

  // isShortlinkValid returns whether a given shortlink is valid.
  isShortlinkValid(shortlink) {
    if (!(typeof shortlink === 'string')) {
      throw new FederatedIssueError('shortlink argument must be a string.');
    }
    return Boolean(shortlink.match(this.shortlinkRe()));
  }

  // shortlinkRe returns the regex used to validate shortlinks.
  shortlinkRe() {
    throw new Error('Not implemented.');
  }

  // toURL returns the URL to this issue.
  toURL() {
    throw new Error('Not implemented.');
  }

  // isOpen returns a Promise that resolves either true or false.
  async isOpen() {
    throw new Error('Not implemented.');
  }
}

/* Class for Google Issue Tracker logic.
 *
 * In order to test this, run the following in the console on an issue detail
 * page that already contains a federated reference to sign in:
 *
 *     gapi.auth2.getAuthInstance().signIn();
 *
 * TODO(monorail:6214): Add authorization button.
 */
export class GoogleIssueTrackerIssue extends FederatedIssue {
  constructor(shortlink) {
    super(shortlink);
    this.issueID = Number(shortlink.substr(2));
  }

  shortlinkRe() {
    return GOOGLE_ISSUE_TRACKER_REGEX;
  }

  toURL() {
    return `https://issuetracker.google.com/issues/${this.issueID}`;
  }

  async isOpen() {
    const userProfile = await loadGapi();
    if (!userProfile) {
      // Fail open.
      return true;
    }

    const res = await this._loadGoogleIssueTrackerIssue(this.issueID);
    if (!res || !res.result) {
      // Fail open.
      return true;
    }

    // Open issues will not have a `resolvedTime`.
    return !Boolean(res.result.resolvedTime);
  }

  get _APIURL() {
    return GOOGLE_ISSUE_TRACKER_API_ROOT + GOOGLE_ISSUE_TRACKER_DISCOVERY_PATH;
  }

  _loadGoogleIssueTrackerIssue(bugID) {
    return new Promise((resolve, reject) => {
      const version = GOOGLE_ISSUE_TRACKER_API_VERSION;
      gapi.client.load(this._APIURL, version, () => {
        const request = gapi.client.corp_issuetracker.issues.get({
          'issueId': bugID,
        });
        request.execute((response) => {
          resolve(response);
        });
      });
    });
  }
}

class FederatedIssueError extends Error {}

// A list of supported tracker classes.
const FEDERATED_TRACKERS = [
  GoogleIssueTrackerIssue,
];
